urlbd77b745af_KokaKiwi_MCLauncher_tgz-pJ8-net_contrapunctus_lzma_BenchmarkJ8/src/com/kokakiwi/mclauncher/api/LauncherAPI.java

```diff
@@ -0,0 +1,335 @@
+            {
+                final Certificate[] certs = ((HttpsURLConnection) connection)
+                        .getServerCertificates();
+                
+                final byte[] bytes = new byte[294];
+                final DataInputStream dis = new DataInputStream(
+                        LauncherAPI.class
+                                .getResourceAsStream("keys/" + keyFile));
+                dis.readFully(bytes);
+                dis.close();
+                
+                final Certificate c = certs[0];
+                final PublicKey pk = c.getPublicKey();
+                final byte[] data = pk.getEncoded();
+                
+                for (int i = 0; i < data.length; i++)
+                {
+                    if (data[i] == bytes[i])
+                    {
+                        continue;
+                    }
+                    throw new RuntimeException("Public key mismatch");
+                }
+            }
+            
+            final DataOutputStream out = new DataOutputStream(
+                    connection.getOutputStream());
+            out.writeBytes(sb.toString());
+            out.flush();
+            out.close();
+            
+            final InputStream in = connection.getInputStream();
+            final BufferedReader reader = new BufferedReader(
+                    new InputStreamReader(in));
+            
+            final StringBuffer response = new StringBuffer();
+            String line;
+            while ((line = reader.readLine()) != null)
+            {
+                response.append(line);
+            }
+            reader.close();
+            
+            result = response.toString();
+        }
+        catch (final Exception e)
+        {
+            e.printStackTrace();
+        }
+        
+        if (connection != null)
+        {
+            connection.disconnect();
+        }
+        
+        return result;
+    }
+    
+    public String getUrl(String dest)
+    {
+        return getUrl(dest, null);
+    }
+    
+    public String getUrl(String dest, String keyFile)
+    {
+        String result = null;
+        HttpURLConnection connection = null;
+        
+        try
+        {
+            final URL url = new URL(dest);
+            connection = (HttpURLConnection) url.openConnection();
+            
+            connection.setRequestMethod("GET");
+            
+            connection.setUseCaches(false);
+            connection.setDoInput(true);
+            connection.setDoOutput(true);
+            
+            connection.connect();
+            
+            if (url.getProtocol().contains("https") && keyFile != null)
+            {
+                final Certificate[] certs = ((HttpsURLConnection) connection)
+                        .getServerCertificates();
+                
+                final byte[] bytes = new byte[294];
+                final DataInputStream dis = new DataInputStream(
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *dis*. The resource is referenced at lines 289, 292. The resource is closed at line 293. There are other execution paths that do not contain closure statements, for example, when DataInputStream.readFully throws an exception. Ensure *dis* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).